src/main.cpp

```diff
@@ -3555,6 +3574,13 @@ bool static LoadBlockIndexDB()
                 pindex->nChainSproutValue = pindex->nSproutValue;
             }
         }
+        // Construct in-memory chain of branch IDs.
+        // Relies on invariant: a block that does not activate a network upgrade
+        // will always be valid under the same consensus rules as its parent.
+        // Activation blocks will have non-zero branch IDs (read from disk).
```

Comment:
I'm nervous about the fact that the unset branch ID (0) is the same as the Sprout branch ID. I think that's a place for bugs to hide. Can we make the unset branch ID -1 instead?

Reply:
Prior to my refactor adding `UPGRADE_SPROUT`, my intention (that I outlined in the ZIP) was that a branch ID of zero meant that no upgrades had been activated. This is exactly equivalent to it being a branch ID for Sprout (which is not actually an upgrade).

If we make the unset branch ID `0xffffffff` instead (as the field is unsigned), it means that we need to:

- Explicitly set the field to that value in `CBlockIndex`.
- Change the check here to check against `0xffffffff` (obviously).
- Explicitly set the branch ID of the genesis block in its index, which means extracting the `if (pindex->pprev)` below into its own wrapping conditional and having an explicit case for the genesis block.
- Ensure that we still don't inadvertently set `BLOCK_ACTIVATES_UPGRADE` for the genesis block (which should be fine, as that is what `IsActivationHeightForAnyUpgrade()` implements).
- Am I forgetting anything?

Reply:
Actually, I think I might rename `UPGRADE_SPROUT` to `BASE_SPROUT` to avoid confusion.